File: app/src/main/java/io/dume/dume/splash/SplashPresenter.java
package io.dume.dume.splash;

import android.app.Activity;
import android.graphics.Color;
import android.os.Build;
import android.util.Log;
import android.view.Window;
import android.view.WindowManager;

import io.dume.dume.auth.AuthGlobalContract;
import io.dume.dume.teacher.homepage.TeacherContract;


public class SplashPresenter implements SplashContract.Presenter {
    SplashContract.View view;
    SplashContract.Model model;
    private static final String TAG = "SplashPresenter";

    public SplashPresenter(SplashContract.View view, SplashContract.Model model) {
        this.view = view;
        this.model = model;
    }

    @Override
    public void init(Activity myActivity) {
        // Status bar :: Transparent
        Window window = myActivity.getWindow();
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS);
            window.addFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS);
            window.setStatusBarColor(Color.TRANSPARENT);
        }
    }

    @Override
    public void enqueue() {
        Log.w(TAG, "enqueue: ");
        model.hasUpdate(new TeacherContract.Model.Listener<Boolean>() {
            @Override
            public void onSuccess(Boolean hasUpdate) {
                Log.w(TAG, "onSuccess: update "+hasUpdate );
                if (hasUpdate) {
                    view.foundUpdates();
                }else {
                    if (model.isUserLoggedIn()) {
                        model.onAccountTypeFound(model.getUser(), new AuthGlobalContract.AccountTypeFoundListener() {
                            @Override
                            public void onStart() {
                                Log.w(TAG, "onStart: ");
                            }

                            @Override
                            public void onTeacherFound() {
                                model.detachListener();
                                view.gotoTeacherActivity();
                                Log.w(TAG, "onTeacherFound: ");

                            }

                            @Override
                            public void onStudentFound() {
                                model.detachListener();
                                view.gotoStudentActivity();
                                Log.w(TAG, "onStudentFound: ");
                            }

                            @Override
                            public void onBootcamp() {
                                model.detachListener();
                                view.gotoTeacherActivity();
                                Log.w(TAG, "onBootcamp: ");
                            }

                            @Override
                            public void onForeignObligation() {
                                model.detachListener();
                                view.gotoForeignObligation();
                                Log.w(TAG, "onForeignObligation: ");
                            }

                            @Override
                            public void onFail(String exeption) {

                            }
                        });

                    } else {
                        view.gotoLoginActivity();
                        Log.w(TAG, "enqueue: login");
                    }
                }
            }

            @Override
            public void onError(String msg) {
                Log.w(TAG, "onError: "+msg );
            }
        });


    }
}

File: app/src/main/java/io/dume/dume/splash/SplashPresenter.kt
package io.dume.dume.splash

import android.content.Context
import io.dume.dume.firstTimeUser.AuthGlobalContract.AccountTypeFoundListener
import io.dume.dume.firstTimeUser.AuthRepository
import io.dume.dume.util.StateManager

class SplashPresenter(var view: SplashContract.View, var repository: AuthRepository) : SplashContract.Presenter {

    override fun enqueue(context: Context) {
        val isFirstTimeUser: Boolean = StateManager.getInstance(context).sharedPreferences().all[StateManager.FIRST_TIME_USER] as? Boolean
                ?: true

        if (isFirstTimeUser) {
            view.gotoForwardFlowActivity()
        } else {
            if (repository.isUserLoggedIn()) {
                repository.onAccountTypeFound(repository.getUser()!!, object : AccountTypeFoundListener {
                    override fun onStart() = run { }
                    override fun onTeacherFound() = view.gotoTeacherActivity()
                    override fun onStudentFound() = view.gotoStudentActivity()
                    override fun onForeignObligation() = view.gotoForeignObligation()
                    override fun onFail(exception: String) = run { view.foundErr(exception) }
                })
            } else view.gotoLoginActivity()
        }
    }

    companion object {
        private const val TAG = "SplashPresenter"
    }

}